Nuget perf telemetry events

jainaashish opened the pull request:
This PR does following:
- Add perf telemetry events for nuget actions like install, update, uninstall, and restore triggered at nuget ui or pmc.
- Refactored existing telemetry code to be more modular and moved to a common place to be used across projects or solution.

Remaining part is to add telemetry events for more detailed steps of these actions.

@rrelyea @joelverhagen @emgarten @alpaix @rohit21agrawal @drewgil @dtivel

dtivel commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/Cmdlets/UpdatePackageCommand.cs:
```diff
@@ -243,10 +288,15 @@ private void PerformPackageUpdatesOrReinstalls()
+
             if (!ShouldContinueDueToDotnetDeprecation(actions, WhatIf.IsPresent))
-            {
+			{
```
Replace tabs with spaces here and below.

dtivel commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/Cmdlets/InstallPackageCommand.cs:
```diff
@@ -49,6 +50,9 @@ protected override void Preprocess()
 
         protected override void ProcessRecordCore()
         {
+            var startTime = DateTime.Now;
```
Should this (and end times too) be `DateTime.UtcNow`?

jainaashish replied:
As Joel commented, it will be DateTimeOffset

dtivel commented on src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -182,38 +183,79 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
+
                     if (!shouldContinue)
                     {
+                        statusMessage = "User cancelled on dotnet depcrecation warning window.";
```
deprecation

dtivel commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/Cmdlets/UninstallPackageCommand.cs:
```diff
@@ -54,12 +55,30 @@ private void Preprocess()
+                startTime,
+                _status,
+                _errorMessage,
+                1,
```
Literal arguments reduce readability:  what does 1 signify?

Options to improve readability include:
1.  Promote the argument to a well-named local:  var packageCount = 1;
2.  Inline comment with the parameter name:  `/*packageCount*/ 1`
3.  Named parameter:  `packageCount: 1`, but unfortunately subsequent arguments would also require named parameters.

jainaashish replied:
Fixed, used already defined variable instead of literal

dtivel commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// operation was successful.
+        /// </summary>
+        Succeed = 1,
```
Succeeded

jainaashish replied:
Fixed

dtivel commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Base class to generate telemetry data for nuget operations like install, udpate or restore.
```
update

dtivel commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+            Duration = duration;
+        }
+
+        public string OperationId { get; set; }
```
Is there any reason why the setters on these properties as well as the setters on all subclasses (ActionStepsTelemetryEvent, RestoreTelemetryEvent) cannot be private?

jainaashish replied:
No reason, earlier i had a different design where i needed those setter but then i changed it. But forgot to remove these setter. But I'll remove these.

dtivel commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+    /// <summary>
+    /// Define different states for nuget operation status.
+    /// </summary>
+    public enum NugetOperationStatus
```
One public type per file?  This feedback applies globally.

jainaashish replied:
I don't think we follow this practice in nuget (at least i haven't seen this) and more so these enum make more sense with these classes to get all information at one place.

dtivel commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/RestoreTelemetryEvent.cs:
```diff
@@ -0,0 +1,51 @@
+        /// <summary>
+        /// Auto restore with nuget restore manager.
+        /// </summary>
+        Implicit =2,
```
Format document (`<space>=<space>`).

dtivel commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/Cmdlets/UpdatePackageCommand.cs:
```diff
@@ -140,10 +161,21 @@ private void PerformPackageUpdatesOrReinstalls()
                     PrimarySourceRepositories,
                     Token);
 
+                if(!actions.Any())
```
Format doc

dtivel commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -362,13 +416,25 @@ private void DisplayOptOutMessage()
                             .ToList();
 
-                        await DependencyGraphRestoreUtility.RestoreAsync(
+                        var restoreSummeries = await DependencyGraphRestoreUtility.RestoreAsync(
```
`restoreSummaries`

drewgillies commented on src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -182,38 +183,79 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
                 var actions = await resolveActionsAsync();
                 var results = GetPreviewResults(actions);
 
+                if(operationType == NugetOperationType.Uninstall)
```
nit: space after if

dtivel commented on src/NuGet.Clients/PackageManagement.VisualStudio/Utility/VsHierarchyUtility.cs:
```diff
@@ -37,6 +37,28 @@ public static IVsHierarchy ToVsHierarchy(Project project)
+            Guid id;
+            if (!hierarchyItem.TryGetProjectId(out id))
+            {
+                id = Guid.Empty;
```
Returning `null` (the absence of a value) or `string.Empty` makes more sense to me here than `{00000000-0000-0000-0000-000000000000}`.

jainaashish replied:
Its just a fallback since in most cases project id will be there. Even though we don't fail without it and just continue which is the existing behavior. I didn't change this default value and will like to continue with it.

drewgillies commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -198,6 +203,11 @@ private void Restore(bool forceRestore, bool showOptOutMessage)
 
+            var stopWatch = new Stopwatch();
+            stopWatch.Start();
+            var statTime = DateTime.Now;
```
startTime

drewgillies commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -540,6 +609,8 @@ var waitDialogFactory
 
             await PackageRestoreManager.RaisePackagesMissingEventForSolutionAsync(solutionDirectory,
                 CancellationToken.None);
+
```
nit: extra line

drewgillies commented on src/NuGet.Core/NuGet.Common/Telemetry/ITelemetrySession.cs:
```diff
@@ -1,8 +1,11 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-namespace NuGet.VisualStudio.Facade.Telemetry
+namespace NuGet.Common
```
Is there a reason for moving this out of the facade namespace? More of a discussion point--I generally like to have namespaces indicate where the class lives. Making it NuGet.Common adds transparency to the facading of this class, which in a sense is removing self-documentation.

jainaashish replied:
Earlier it was part of VisualStudio facade project in client sln but now i've moved into common project in core sln. And the reason to move it there is to have more flexibility and ability to emit telemetry events across solutions.

dtivel commented on src/NuGet.Clients/VisualStudio.Facade/Telemetry/TelemetrySession.cs:
```diff
@@ -2,13 +2,18 @@
 {
     public class TelemetrySession : ITelemetrySession
     {
+        public static TelemetrySession Instance = new TelemetrySession();
```
Change to a read-only property.

jainaashish replied:
Updated

drewgillies commented on test/NuGet.Core.Tests/NuGet.Common.Test/Telemetry/ActionsTelemetryServiceTests.cs:
```diff
@@ -0,0 +1,129 @@
+                source,
+                DateTime.Now.AddSeconds(-3),
+                NugetOperationStatus.Succeed,
+                string.Empty,
```
param names on some of these less self-explaining args would be helpful

jainaashish replied:
Done

joelverhagen commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:
```diff
@@ -49,6 +49,13 @@ public abstract class NuGetPowerShellBaseCommand : PSCmdlet, IPSNuGetProjectCont
         private readonly IHttpClientEvents _httpClientEvents;
 
+        // default stopWatch to calculate total time elapsed for an action
+        protected Stopwatch ActionStopWatch = new Stopwatch();
```
> ActionStopWatch [](start = 28, length = 15)

Why share this instance? Couldn't this be declared in a narrower scope?

jainaashish replied:
No, because at many times, we go back and forth from command specific class to base class and have user confirmation dialogs. Since we want to only calculate actual time taken by nuget to process an action and avoiding noise like time to react to a ui dialog, etc... so we need this at this level, otherwise it will be very complicated to add/subtract time duration.

joelverhagen commented on src/NuGet.Clients/PackageManagement.UI/Utility/TelemetryUtility.cs:
```diff
@@ -0,0 +1,60 @@
+        /// <param name="endTime"></param>
+        /// <param name="duration"></param>
+        /// <returns></returns>
+        public static ActionsTelemetryEvent GetActionTelemetryEvent(
```
> GetActionTelemetryEvent [](start = 44, length = 23)

Could this method emit as well?

jainaashish replied:
It could but then the purpose of this method will change besides i'll also need to pass telemetry service instance to emit event. Here i just wanted to construct the event object to avoid duplicate code.

joelverhagen commented on src/NuGet.Clients/PackageManagement.UI/Utility/TelemetryUtility.cs:
```diff
@@ -0,0 +1,60 @@
+            IEnumerable<NuGetProject> projects,
+            NugetOperationType operationType,
+            OperationSource source,
+            DateTime startTime,
```
> DateTime [](start = 12, length = 8)

Use DateTimeOffset

jainaashish replied:
Fixed

joelverhagen commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionsTelemetryEvent.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <summary>
+    /// Define nuget operation type values.
+    /// </summary>
+    public enum NugetOperationType
```
> NugetOperationType [](start = 16, length = 18)

Uppercase G in NuGet

jainaashish replied:
Fixed

joelverhagen commented on src/NuGet.Core/NuGet.Common/Telemetry/Constants.cs:
```diff
@@ -0,0 +1,47 @@
+    /// <summary>
+    /// This class contains telemetry events name and properties name.
+    /// </summary>
+    public static class Constants
```
> Constants [](start = 24, length = 9)

TelemetryConstants

jainaashish replied:
Fixed

joelverhagen commented on src/NuGet.Core/NuGet.Common/Telemetry/Constants.cs:
```diff
@@ -0,0 +1,47 @@
+        public const string ProjectIdPropertyName = PropertyNamePrefix + "ProjectId";
+
+        // nuget telemetry event names
+        public const string ProjectInformationEventName = EventNamePrefix + "ProjectInformation";
```
> const [](start = 15, length = 5)

public consts should be static readonly (NuGet convention)

jainaashish replied:
Fixed

joelverhagen commented on src/NuGet.Core/NuGet.Common/Telemetry/NuGetTelemetryService.cs:
```diff
@@ -0,0 +1,38 @@
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Base class for nuget telemetry services
```
> Base [](start = 8, length = 4)

I don't see what inherits this. What's the purpose of this class? Can't the caller just new up a telemetry event on their own?

jainaashish replied:
I was thinking to have a common place to emit telemetry event... But after a second thought, I think you're right we dont actually need this and specific telemetry service class will be able to just call into telemetry session and pass telemetry event.

joelverhagen commented on src/NuGet.Clients/PackageManagement.VisualStudio/Utility/VsHierarchyUtility.cs:
```diff
@@ -37,6 +37,28 @@ public static IVsHierarchy ToVsHierarchy(Project project)
+            return new VsHierarchyItem(ToVsHierarchy(project));
+        }
+
+        public static string GetProjectId(Project project)
```
> GetProjectId [](start = 29, length = 12)

We need to be careful about changing this. It's possible downstream is not resilient to consuming non-GUID values. Please investigate.

jainaashish replied:
Its not being used any where, we use it here only to attach these project id to nuget project.

emgarten replied:
Do we really need the project id or does this just need a unique id? If possible this should just use the object hash code that relates to the memory reference or some other unique randomly generated guid. Getting the ProjectId impacts perf.

jainaashish replied:
But ProjectId is the only thing could be used to related to other telemetry tables in Kusto to get more details. So yeah we need project id here.

joelverhagen commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -198,6 +203,11 @@ private void Restore(bool forceRestore, bool showOptOutMessage)
             _displayRestoreSummary = false;
             _hasOptOutBeenShown = !showOptOutMessage;
 
+            var stopWatch = new Stopwatch();
```
> Stopwatch [](start = 32, length = 9)

nit: Stopwatch.StartNew

joelverhagen commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionStepsTelemetryEvent.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public string OperationId { get; }
+
+        public string StepName { get; }
```
> StepName [](start = 22, length = 8)

This could be an enum, right?

jainaashish replied:
Not really since its for our use n debugging purpose so i would like to keep it as string and have sufficient description in each step name. Also the fact that there will be different steps with different actions.

joelverhagen commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -490,6 +558,7 @@ var waitDialogFactory
                     return;
                 }
 
+                _packageCount += packages.Count;
```
> _packageCount [](start = 16, length = 13)

Presumable we need instance state here because these methods are void. Could we avoid this by making RestorePackagesOrCheckForMissingPackagesAsync (and the other restore method) return a RestoreTelemetryData (naming?) private type?

jainaashish replied:
`Presumable we need instance state here because these methods are void`
I'm not sure what exactly you mean by this? But ideally i would have also avoided updating instance variable like this but then it's not really updating instance state which could impact outside. Since _packageCount is a private variable n also being updated in private api

emgarten replied:
This class is a singleton that lives forever in VS. Keeping a long error string under _errors and _packageCount on the class means that this will also stay around as long as VS is open. There is no reason to persist this.

jainaashish replied:
Removed _errors.

joelverhagen commented on src/NuGet.Clients/VisualStudio.Facade/project.json:
```diff
@@ -1,5 +1,6 @@
 ﻿{
   "dependencies": {
+    "NuGet.Common": "3.6.0-*",
```
> Common [](start = 11, length = 6)

How have we avoid this dependency until now? Are all of the facade types fully declared in this assembly up until now?

jainaashish replied:
Until now we didn't need to access Nuget.Common package in facade project, this is the new dependency.

jainaashish commented:
Fixed PR comments and added perf events for granular level. Will add more tests to trigger these events for actual install/ update or restore action.

emgarten commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/PackageActionBaseCommand.cs:
```diff
@@ -108,11 +109,25 @@ protected override void ProcessRecordCore()
+                }
+                else
+                {
+                    _packageCount = actions.Select(action => action.PackageIdentity.Id).Distinct().Count();
```
Distinct(StringComparer.OrdinalIgnoreCase) should the actions somehow have different cases

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -23,6 +24,7 @@ public class UIActionEngine
     {
         private readonly ISourceRepositoryProvider _sourceProvider;
         private readonly NuGetPackageManager _packageManager;
+        private Stopwatch ActionStopWatch = new Stopwatch();
```
readonly

emgarten commented:
🔥 This should not go into dev. The cost of maintaining all of these data points is very high with little benefit since the resulting data from many of these conveys just a time and no information on what occurred.

The logging here is equivalent to perf tracing by method, this very tightly couples the current implementation of the client code to external reports and makes it impossible to refactor or make major changes and still capture the same information. 

Telemetry should provide data that can be used to answer a high level question that can be acted on, profiling the current implementation should be done on dev machines, not on user machines. Comparing the inputs with the profiling results is the only way to make sense of the numbers.

### Tool restore example / refactoring
Tool restores have already changed in dev to run in parallel outside of the project but the telemetry for tool restore perf is done by just timing how long the tool restore method takes. Returning a number that is comparable is impossible because 1) We now de-duplicate tool restores  2) Running in parallel means they will be competing for CPU and bandwidth. If you tried to compare the results from xproj tool restore to .NETCore tool restores it would lead you to wrong conclusions such as 1) There are far less tools being restored in .NETCore, 2) tool restores take 1000x longer (since they are de-duped and there will no longer be Tools x Projects noops in a solution restore). This telemetry provides no way to get to the correct conclusion here from what I see. 

There are few advantages here, only major blockers to refactoring things when someone now has to rearrange all of this telemetry code and fix all the tests around this.

### Graph generation example
Each restore graph generated is measured and tracked by the project id. This will give you a time which is equivalent to a random number without knowing any of the following

1. Was this is a noop?
1. Did this fail completely?
1. Did downloads occur during the restore? What type of source?
1. Was everything already cached from the graph before it?
1. Is this a RID or RIDless graph?
1. Are other restores going on in parallel?
1. Are there 0, 3, or 20934324 packages in this graph?

Graph generation time and the number of times it is executed would also change dramatically if we get to caching/reusing graphs for solution restores. There is no way to maintain this as useful if anything is refactored, comparing numbers between releases will be the same as comparing numbers between different scenarios.

### Suggestions

1. Reduce the telemetry points down to 2 or 3 critical pieces of information that we can maintain for several releases.
1. Do not tie telemetry to the implementation of the code. Measure events that can easily be understood by someone who has never seen the code, and possibly that could be applied to any package manager. Anything else will likely be unusable in the next release.
1. The business logic for telemetry should not be in the product code to reduce the chance of someone breaking it when making other changes. Create a **static** telemetry service and make simple method calls to pass the data into it. All business logic should live inside the telemetry service, the client code should have only a single line of new code at the start and end, or a single line reporting on the result.
1. Every line of telemetry code outside of the telemetry service **must** contain a paragraph describing what is measuring, how it is reported, and the goal so that when refactoring or fixing issues in that area it can be possible to keep the original intent of the telemetry or to decide if it is no longer valid. Without this it will be trivial to break or regress what is actually being captured/measured and it will not be noticed until someone tries to use the data long after the release.

### Explore the results
Try running this several times on ~50 solutions and logging the results. It should give you a good idea for what this is going to output, then see if the information can be used.

//cc @harikmenon @rrelyea

emgarten commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionsTelemetryService.cs:
```diff
@@ -0,0 +1,75 @@
+    /// </summary>
+    public class ActionsTelemetryService
+    {
+        protected readonly ITelemetrySession telemetrySession;
```
Remove all of the telemetry classes from NuGet.Common. It does not make sense for MSBuild, CLI, and all 3rd party users of the libraries to take on this dependency for Visual Studio only telemetry.

Having it in NuGet.Common will very likely lead to users listening to the events for other purposes and being broken when they change.

emgarten commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -76,6 +80,9 @@ internal class OnBuildPackageRestorer : ILogger
         private bool _hasOptOutBeenShown;
+        private string _errorMessage = string.Empty;
+        private NuGetOperationStatus _status;
+        private int _packageCount;
```
Having telemetry related fields on classes like this is very confusing, remove them and store this data outside of the class.

There is no indication of what this is for, or why on build package restorer should track the package count across all restores on the singleton instance of this.

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/VSNuGetProjectFactory.cs:
```diff
@@ -54,13 +54,16 @@ public NuGetProject CreateNuGetProject(EnvDTEProject envDTEProject, INuGetProjec
                     envDTEProject.Name,
-                    EnvDTEProjectUtility.GetCustomUniqueName(envDTEProject));
+                    EnvDTEProjectUtility.GetCustomUniqueName(envDTEProject),
+                    projectId);
```
These projects won't exist soon and all actions for these projects simply pass the id/version to an interface. There is no indication of what happened with it.

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/VSNuGetProjectFactory.cs:
```diff
@@ -54,13 +54,16 @@ public NuGetProject CreateNuGetProject(EnvDTEProject envDTEProject, INuGetProjec
 
             NuGetProject result = null;
 
+            var projectId = VsHierarchyUtility.GetProjectId(envDTEProject);
```
Changes such as this have a major impact on performance and can often cause unexpected errors for unknown project types. For example the change in 3.4.0 to just check for the project path up front in this same area lead to several serious bugs. Avoid doing this if possible.

jainaashish replied:
We're already doing this as part of our project telemetry events... I haven't added anything new just moved it to right place so that it could be used across nuget project

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs:
```diff
@@ -51,6 +51,8 @@ public class RestoreArgs
 
         public int? LockFileVersion { get; set; }
 
+        public ActionsTelemetryService NugetTelemetryService { get; set; }
```
Remove this from NuGet.Commands, all Visual Studio related restore data can be gathered from the summaries returned to Visual Studio.

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2597,7 +2688,7 @@ private Task OpenReadmeFile(NuGetProject nuGetProject, INuGetProjectContext nuGe
                     if (!string.IsNullOrEmpty(packageFolderPath))
                     {
-                        readmeFilePath = Path.Combine(packageFolderPath, Constants.ReadmeFileName);
+                        readmeFilePath = Path.Combine(packageFolderPath, ProjectManagement.Constants.ReadmeFileName);
```
Why did this change?

jainaashish replied:
Not needed this change. Reverted

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2446,6 +2519,12 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
 
+                stopWatch.Stop();
+                EmitTelemetryEvent(
+                    string.Format(TelemetryConstants.WritingLockFileStepName, projectName),
```
Between the start and the stop of the timer here one of two things could happen:
1. An already computed lock file could be written to disk, this is ~200ms of time and probably never above 1000ms
2. The project could be restored, the http cache could expire and floating packages would need to be reevaluated, it could then download 50 packages, extract them to disk, compute the restore graph for 10+ frameworks and RIDs. This could take 1 minute to 5 hours depending on your internet speed.

Both of these things will be logged to the same telemetry point of: Writing the assets file.  There is no way to use this data!

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:
```diff
@@ -63,11 +63,12 @@ public static class DependencyGraphRestoreUtility
         }
 
-        public static async Task RestoreAsync(
+        public static async Task<IReadOnlyList<RestoreSummary>> RestoreAsync(
```
It does not make sense for a telemetry change to both need all the summary data passed back up, and the telemetry service passed down in this method.

jainaashish replied:
Not passing telemetry service anymore.

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -55,6 +55,8 @@ public class NuGetPackageManager
 
         public IInstallationCompatibility InstallationCompatibility { get; set; }
 
+        public ActionsTelemetryService NugetActionsTelemetryService { get; set; }
```
Remove this. If the telemetry service needs to be passed around on every class, that indicates that it should be referenced from a static instance somewhere.

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -647,8 +649,8 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
-            
-            var buildIntegratedProjects = nuGetProjects.OfType<BuildIntegratedNuGetProject>().ToList();
+
+            var buildIntegratedProjects = nuGetProjects.Where(project => project is INuGetIntegratedProject);
```
Why did this change?

jainaashish replied:
Because it should follow the same behavior for Dotnet core projects as it does for UWP projects. And both these types extends from INuGetIntegratedProject

emgarten commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1123,6 +1147,11 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
 
+                stopWatch.Stop();
+                EmitTelemetryEvent(
+                    string.Format(TelemetryConstants.ResolvedActionsStepName, projectName),
```
These steps are extremely granular and specific and appear to be trying to test the product on customer machines by looking for unusually long times here.

I don't see how the times logged here could be useful for a report viewed outside of the nuget team. And I don't see how the nuget team can use the data collected here. Please provide some details on this or a spec with sample output that is expected and a plan for how it could be used.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -30,6 +30,7 @@ public class RestoreCommand
     {
         private readonly ILogger _logger;
         private readonly RestoreRequest _request;
+        private readonly ActionsTelemetryService _telemetryService;
```
This is already on the _request object right above it, why would this possibly need to be here?

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs:
```diff
@@ -155,5 +155,10 @@ public class RestoreRequest
+        /// <summary>
+        /// This will be used to trigger restore perf telemetry events at granular level of restore.
+        /// </summary>
+        public ActionsTelemetryService NugetTelemetryService { get; set; }
```
Remove this everywhere from RestoreCommand

emgarten commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -362,13 +431,26 @@ private void DisplayOptOutMessage()
+                        if (isRestoreFailed)
+                        {
+                            _status = NuGetOperationStatus.Failed;
+                            _errorMessage = string.Join(",", restoreSummaries.SelectMany(summary => summary.Errors));
```
:fire: THIS WILL LOG PERSONALLY IDENTIFIABLE INFORMATION

Errors often contain
- Path to the global package folder which contains the user name
- Internal package names
- Internal source urls/paths when download failures occur
- Error responses from proxies
- Credential related error messages

emgarten commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -362,13 +431,26 @@ private void DisplayOptOutMessage()
+                            referenceContext,
+                            _telemetryService);
+
+                        _packageCount += restoreSummaries.Select(summary => summary.InstallCount).Sum();
```
If there are 10 identical projects with 10 packages this will return 100. Which is not a useful number here.

emgarten commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -216,13 +230,20 @@ private void Restore(bool forceRestore, bool showOptOutMessage)
                     // Note that projects that are not supported by NuGet, will not show up in this list
                     var projects = SolutionManager.GetNuGetProjects().ToList();
 
+                    var source = showOptOutMessage == true ? RestoreOperationSource.OnBuild : RestoreOperationSource.Explicit;
```
Why use showOptOutMessage instead of force, which is the what is set when the user explicitly performs a restore?

emgarten commented on src/NuGet.Clients/VsExtension/VsEvents/OnBuildPackageRestorer.cs:
```diff
@@ -216,13 +230,20 @@ private void Restore(bool forceRestore, bool showOptOutMessage)
+                        stopWatch.Stop();
+                        _telemetryService?.EmitActionStepsEvent(TelemetryConstants.RestorePackagesConfigStepName,
+                            stopWatch.Elapsed.TotalSeconds);
+                        stopWatch.Start();
```
The pattern of starting the timer for whoever is next feels very fragile and error prone. This should be handled by the telemetry service, and only started explicitly before the part you want to measure.

emgarten commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+
+        public DateTimeOffset EndTime { get; }
+
+        public double Duration { get; }
```
What is the reason for logging StartTime, EndTime, AND duration? When are all needed together? 

I would expect just one TimeSpan here. If the start/end are needed the time the object was created could be logged as the end time, and everything else figured out from there internally.

emgarten commented on src/NuGet.Core/NuGet.ProjectManagement/Projects/MSBuildNuGetProject.cs:
```diff
@@ -73,7 +73,8 @@ public class MSBuildNuGetProject : NuGetProject, IDependencyGraphProject
             string folderNuGetProjectPath,
-            string packagesConfigFolderPath)
+            string packagesConfigFolderPath,
+            string projectId = null)
```
Do not use optional parameters, it's part of the coding guidelines.

It is likely that changing this involves tons of test changes across the solution, I don't see why this change is needed though. Do not add this parameter.

jainaashish replied:
Fixed

emgarten commented on src/NuGet.Core/NuGet.Common/Telemetry/Performance/ActionEventBase.cs:
```diff
@@ -0,0 +1,70 @@
+    /// <summary>
+    /// Base class to generate telemetry data for nuget operations like install, update or restore.
+    /// </summary>
+    public abstract class ActionEventBase
```
Do not use a generic name for this, it is specific to telemetry.

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/Telemetry/ProjectTelemetryEvent.cs:
```diff
@@ -7,7 +7,7 @@ namespace NuGet.PackageManagement.Telemetry
     public abstract class ProjectTelemetryEvent
     {
-        public ProjectTelemetryEvent(string nuGetVersion, Guid projectId)
+        public ProjectTelemetryEvent(string nuGetVersion, string projectId)
```
Why are there several changes related to first classing the project id guid at the project level, yet for these new changes the project name is used all over and passed in here as the project id?

The project name seems like it would be considered personally identifiable and unexpectedly give away details on a non-public project. To avoid that this should really stay as a Guid.

jainaashish replied:
It's the project id only. Earlier we use to pass the Guid and while emitting event, convert it to string. But now we just take project id as string.

emgarten commented:
A few more things I'm concerned about here

* Error messages are being logged here which will contain sensitive information.
* Project names are logged instead of the project Guid which will contain sensitive information.
* The changes are very intrusive at every level of the client here. This should be limited to PackageManagement.VisualStudio as it was before. Collecting and attempting to log telemetry data for Visual Studio outside of Visual Studio scenarios such as in the CLI and MSBuild does not make sense. Currently this code lives in NuGet.Common which is the absolute lowest level.
* Performance will be an issue here, along with increased memory usage since things are being stored indefinately to classes that are singletons in MEF.
* There are no comments explaining any of this where it is logged which makes it very hard for external users to work in the code and contribute. The names of most collection variables here do not imply it is used for telemetry either.
* The events here are extremely granular in certain spots, such as logging the time it takes to write an already computed lock file to disk. It gives the impression that this is putting test code onto user's machines with this and do not have a plan for using this telemetry.

jainaashish commented:
PR is updated with as per our recent discussion around this perf measurement.

@emgarten

jainaashish commented:
Merged ca0bb4588a366f07af31e435f301b72e6ae7a46d

emgarten commented:
@jainaashish Were these changes signed off on?

emgarten commented:
This change also does not look like what we talked about.

jainaashish commented:
Lets discuss offline.

Commit:
Fixed PR comments

Updated design to avoid updating NugetPackageManager's properties for telemetry service.

jainaashish commented:
Updated design to avoid updating nugetpackagemanager property from outside or using singleton to emit granular level telemetry events. Both these solutions weren't appropriate and adding further complexity to already complex npm.

Resolved it through INugetProjectContext which seems right thing given the context and keep this telemetry stuff independent from packagemanager. It also makes it easy for testing unlike singleton or static class design.

@emgarten @joelverhagen @alpaix please review and let me know if you've any concern, otherwise I'll go ahead and check it in by EOD.
